### PR TITLE
Lease and election management, take two

### PR DESCRIPTION
**Pull request by nightkr**
This builds on #791, but tries to be more futures-native. In particular:

1. It uses regular future cancellation when on failures
2. It uses future multiplexing (`futures::future::select`) rather than spawning new Tokio tasks for managing renewal

TODO:

- [ ] Fix deadlock where leases are never stolen, even after the original owner's lease has expired
- [ ] Don't start running before watchdog is configured
- [ ] Reuse the same watcher for all steps
- [ ] Retries
- [ ] General cleanup and refactoring
- [ ] Clean up error types
- [ ] Docs
- [ ] Tests

### CODE REVIEW

File: examples/elector.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    } else {
+        kube::Api::<Lease>::default_namespaced(kube)
+    };
+    let elector = Elector::new(leases, &lease, &instance, lease_duration_secs);
+    elector
+        .run(async {
+            tracing::info!("acquired lease, press <ctrl+c> to release...");
+            ctrl_c().await
+        })
+        .await??;
```
Comment (clux) [inline examples/elector.rs:44]: I like this interface a lot more. :+1:

File: kube-runtime/src/lease.rs [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+    fn state(&self, lease: &LeaseSpec) -> LeaseState {
+        let lease_duration = Duration::seconds(lease.lease_duration_seconds.unwrap_or(0).into());
+        let last_renewal = lease.renew_time.as_ref().map_or(MIN_DATETIME, |dt| dt.0);
+
+        match &lease.holder_identity {
+            None => LeaseState::Unheld,
+            Some(holder) if holder == &self.identity => LeaseState::HeldBySelf {
+                expires_at: last_renewal + lease_duration,
+                renew_at: last_renewal + lease_duration / 2,
```
Comment (clux) [inline kube-runtime/src/lease.rs:214]: so this is a renewal at half of the lease duration. this makes sense as an initial default, but this is configurable in client-go. I think they by default they waited 80% of the duration before renewing IIRC.

EDIT: they also have a bunch of [jitter logic on their retries](https://github.com/kubernetes/client-go/blob/8e46da3fd12b54a2d1726fc9a9438ff8fc38494c/tools/leaderelection/leaderelection.go#L249-L260) :thinking:

**Review by clux — COMMENTED** [review]
this looks much cleaner and more what i am used to. do you think the best way to go is to just kill the app by default when a lease expires? this seems to be what [client-go does in the example](https://github.com/kubernetes/client-go/blob/master/examples/leader-election/main.go#L134-L154). i was originally thinking we could explore a **guard** type pattern that we run users code inside (so that we would hold off until a lease has been reacquired), but maybe that's not very smart.

**Comment by teozkr** [comment]
> i was originally thinking we could explore a guard type pattern that we run users code inside (so that we would hold off until a lease has been reacquired), but maybe that's not very smart.

We can sort of do that with the current `run` interface (just stop polling the future while paused). The big caveats there are:

1. Not all use cases are reentrant. Resuming execution after regaining the lease (if it has been taken by another client in the meantime) could lead to unsafety.
2. Subtasks spawned onto the executor (for example, via `tokio::spawn`) would still continue running normally. For `Controller` this means that we would stop listening for new events or scheduling new reconciliations, but existing reconciliations would still continue running normally.

File: kube-runtime/src/lease.rs (hunk 2 above)
Comment (teozkr) [reply kube-runtime/src/lease.rs:214]: Definitely open for thoughts on this, this is fairly placeholdery. We also need to have some clock skew compensation between `LeaseState::HeldBySelf`'s `expires_at` and `HeldByOther`'s. Keeping this discussion open for now.

File: examples/elector.rs (hunk 1 above)
Comment (teozkr) [reply examples/elector.rs:44]: Good to hear!

**Comment by teozkr** [comment]
I'm putting this on hold for now, the corner cases are driving me slightly insane. I still want to have this eventually, but I'm not sure this is quite it either.

**Comment by xrl** [comment]
Hey, this is great!

**Comment by clux** [comment]
Closing this along with all other abandoned leader election PRs for now as per #485.

**Pull request closed without merging**